### DISCUSSION TITLE
The best practice for using the blocking mutex in the asynchronous context

### DISCUSSION THREAD

**Discussion by vvvviiv** [tokio-rs/tokio#6785] [Q&A] [answered]
As the doc of `tokio::sync::Mutex` said:
https://github.com/tokio-rs/tokio/blob/5ea3c63d7e5fef97a8e7abf329d9d00cdbc90661/tokio/src/sync/mutex.rs#L26-L27

And since `Tokio` provides `block_in_place`:
https://github.com/tokio-rs/tokio/blob/5ea3c63d7e5fef97a8e7abf329d9d00cdbc90661/tokio/src/task/blocking.rs#L4-L5
which is suitable for synchronous contexts:
https://github.com/tokio-rs/tokio/blob/5ea3c63d7e5fef97a8e7abf329d9d00cdbc90661/tokio/src/task/blocking.rs#L23-L25

We have two ways to lock a blocking mutex:

1. Call the blocking `lock` method directly:

    ```rust
    #[tokio::main]
    async fn main() {
        let mutex = std::sync::Mutex::new(String::new());

        *mutex.lock().unwrap() = "hello world".to_string();
    }
    ```

2. Call the blocking `lock` method inside `block_in_place`:

    ```rust
    #[tokio::main]
    async fn main() {
        let mutex = std::sync::Mutex::new(String::new());

        tokio::task::block_in_place(|| {
            *mutex.lock().unwrap() = "hello world".to_string();
        });
    }
    ```

The second approach may be better as it avoids blocking the executor.

Do I understand this correctly? Thanks for your help.

**Comment by Darksonn** [comment, marked as answer]
When using a blocking mutex, you must ensure that the call to `lock` takes a short amount of time. And if it does that, it's not really blocking.

Using `block_in_place` is a bad idea. The way it works is that a new worker thread is spawned in the background. It's not a magic bullet and it can be really expensive. If locking the mutex takes a short amount of time, then it is unnecessary. If locking the mutex takes a long time, you should use `tokio::sync::Mutex` instead of `block_in_place`.

**Reply by vvvviiv** [reply]
That's really helpful, thank you!
